ska_dbi/sqsh.py

```diff
@@ -0,0 +1,168 @@
+
+    :rtype: Sqsh object
+    """
+    def __init__(
+        self,
+        server=None,
+        user=None,
+        passwd=None,
+        database=None,
+        sqshrc=None,
+        authdir="/proj/sot/ska/data/aspect_authorization",
+        **kwargs,
+    ):
+
+        self.server = server or DEFAULT_CONFIG['sybase'].get("server")
+        self.user = user or DEFAULT_CONFIG['sybase'].get("user")
+        self.database = database or DEFAULT_CONFIG['sybase'].get("database")
+        self.passwd = passwd
+        self.sqshrc = sqshrc
+
+        if not Path(SYBASE).exists():
+            raise RuntimeError(f"SYBASE does not exist: {SYBASE}")
+        if not Path(SQSH_BIN).exists():
+            raise RuntimeError(f"sqsh.bin does not exist: {SQSH_BIN}")
+
+        if self.passwd is None and self.sqshrc is None:
+            sqshrc_file = (
+                Path(authdir) / f"sqsh-{self.server}-{self.database}-{self.user}"
+            )
+            if sqshrc_file.exists():
+                self.sqshrc = sqshrc_file
+            else:
+                raise NoPasswordError(
+                    f"None supplied and unable to read inferred sqshrc file {sqshrc_file}"
+                )
+
+    def __enter__(self):
+        """Context manager enter runtime context.  No action required, just return self."""
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """
+        Context manager exit run time context.
+        """
+        pass
+
+
+    def fetch(self, query):
+        """
+        Execute a query and return all returned sql rows a list of lines.
+
+        Note that this doesn't do anything fancy with the types of the return columns.
+        Python Sybase, for example, handled Sybase datetime columns as Python datetimes types
+        and here they are just strings.
+
+        Parameters
+        ----------
+        query : str
+            SQL query to execute (select statements are expected for our use cases)
+
+        Returns
+        -------
+        list of str
+        """
+        envs = ENV_VARS
+        if self.sqshrc is not None:
+            envs = envs + [f"SQSHRC={self.sqshrc}"]
+        cmd = [
+            SQSH_BIN,
+            "-X",
+            "-S",
+            self.server,
+            "-U",
+            self.user,
+            "-D",
+            self.database,
+            "-m",
+            "csv",
+            "-C",
+            query,
+        ]
+        if self.passwd is not None:
+            cmd += ["-P", self.passwd]
+        cmd = ["env"] + envs + cmd
+
+        proc = subprocess.Popen(
+            cmd, stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE
+        )
+        stdout, stderr = proc.communicate()
+        if proc.returncode != 0:
+            raise RuntimeError(
+                "sqsh failed with return code %d: %s" % (proc.returncode, stderr)
+            )
+
+        outlines = stdout.decode().splitlines()
```

Comment:
This looks pretty close to a single call to `subprocess.check_output()`. Thoughts?

Reply:
That's fine.  I think with communicate() or check_output() if we don't do anything fancy, this would give error if the sqsh.bin wasn't available for example but wouldn't have useful output if the query was poorly constructed.